test/functional/p2p_orphan_handling.py

```diff
@@ -0,0 +1,416 @@
+        peer.send_message(msg_notfound(vec=[CInv(MSG_WITNESS_TX, int(txid_conf_old, 16))]))
+        peer.send_and_ping(msg_tx(missing_tx["tx"]))
+        peer.sync_with_ping()
+        assert_equal(node.getmempoolentry(orphan["txid"])["ancestorcount"], 3)
+
+    @cleanup
+    def test_orphans_overlapping_parents(self):
+        node = self.nodes[0]
+        # In the process of relaying inflight_parent_AB
+        peer_txrequest = node.add_p2p_connection(PeerTxRelayer())
+        # Sends the orphans
+        peer_orphans = node.add_p2p_connection(PeerTxRelayer())
+
+        confirmed_utxos = [self.wallet_nonsegwit.get_utxo() for _ in range(4)]
+        assert all([utxo["confirmations"] > 0 for utxo in confirmed_utxos])
+        self.log.info("Test handling of multiple orphans with missing parents that are already being requested")
+        # Parent of child_A only
+        missing_parent_A = self.wallet_nonsegwit.create_self_transfer(utxo_to_spend=confirmed_utxos[0])
+        # Parents of child_A and child_B
+        missing_parent_AB = self.wallet_nonsegwit.create_self_transfer(utxo_to_spend=confirmed_utxos[1])
+        inflight_parent_AB = self.wallet_nonsegwit.create_self_transfer(utxo_to_spend=confirmed_utxos[2])
+        # Parent of child_B only
+        missing_parent_B = self.wallet_nonsegwit.create_self_transfer(utxo_to_spend=confirmed_utxos[3])
+        child_A = self.wallet_nonsegwit.create_self_transfer_multi(
+            utxos_to_spend=[missing_parent_A["new_utxo"], missing_parent_AB["new_utxo"], inflight_parent_AB["new_utxo"]]
+        )
+        child_B = self.wallet_nonsegwit.create_self_transfer_multi(
+            utxos_to_spend=[missing_parent_B["new_utxo"], missing_parent_AB["new_utxo"], inflight_parent_AB["new_utxo"]]
+        )
+
+        # The wtxid and txid need to be the same for the node to recognize that the missing input
+        # and in-flight request for inflight_parent_AB are the same transaction.
+        assert_equal(inflight_parent_AB["txid"], inflight_parent_AB["tx"].getwtxid())
+
+        # Announce inflight_parent_AB and wait for getdata
+        peer_txrequest.send_and_ping(msg_inv([CInv(t=MSG_WTX, h=int(inflight_parent_AB["tx"].getwtxid(), 16))]))
+        self.nodes[0].bumpmocktime(NONPREF_PEER_TX_DELAY)
+        peer_txrequest.wait_for_getdata([int(inflight_parent_AB["tx"].getwtxid(), 16)])
+
+        self.log.info("Test that the node does not request a parent if it has an in-flight txrequest")
+        # Relay orphan child_A
+        self.relay_transaction(peer_orphans, child_A["tx"])
+        self.nodes[0].bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        # There are 3 missing parents. missing_parent_A and missing_parent_AB should be requested.
+        # But inflight_parent_AB should not, because there is already an in-flight request for it.
+        peer_orphans.wait_for_parent_requests([int(missing_parent_A["txid"], 16), int(missing_parent_AB["txid"], 16)])
```

Comment:
nit: I was a bit irritated that that `inflight_parent_AB` not being requested isn't checked here but I see it follows below. For expressiveness you could consider duplicating the check here or adding a small comment.